Add static dispatch mode to reduce mobile code size

li-roy opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#22335 Add static dispatch mode to reduce mobile code size**

As we discussed, this will allow the linker to remove unused operators automatically. 

Differential Revision: [D16048264](https://our.internmc.facebook.com/intern/diff/D16048264)

Commit:
Add static dispatch mode to reduce mobile code size

Commit:
Update on "Add static dispatch mode to reduce mobile code size"

Add static dispatch mode to reduce mobile code size

gh-metadata: pytorch pytorch 22335 gh/li-roy/36/head

Commit:
Update on "Add static dispatch mode to reduce mobile code size"

Add static dispatch mode to reduce mobile code size

gh-metadata: pytorch pytorch 22335 gh/li-roy/36/head

Commit:
Update on "Add static dispatch mode to reduce mobile code size"

Add static dispatch mode to reduce mobile code size

gh-metadata: pytorch pytorch 22335 gh/li-roy/36/head

smessmer commented on aten/src/ATen/core/TensorMethods.h:
```diff
@@ -69,1622 +75,4624 @@ inline void Tensor::set_data(Tensor new_data) {
+#endif
 }
 inline Tensor Tensor::add(const Tensor & other, Scalar alpha) const {
+#ifdef USE_STATIC_DISPATCH
```
Actually, seeing this and it's similarity to the operator registration API, I got a maybe even better idea. All we need is to make sure that the `CPUType::add` kernel is present if `Tensor::add` is called. If we know it's present, we don't even need a separate static dispatcher and can just go through c10. So what if we dispatch through c10 but if it's a mobile build, we add a

    #if defined(USE_STATIC_DISPATCH)
    static auto registry = torch::RegisterOperators()...
    #endif

(or the equivalent for the intermediate dispatcher) at the top of the tensor method? Do you think that would work?

I'm not sure which approach is better but I'd like to hear your opinion.

li-roy replied:
I feel like it's hard to factor it out because we'd be adding another layer of indirection, and in this case we'd be generalizing two cases that aren't all that similar. The static case needs the functions for each cpu backend, and the dispatcher case needs a schema/op name. Not sure if it's worth.

I think your approach of registering inside the method call will work. I think the main problem with that approach is that we need to make dispatches in `TensorMethods.h` and `Functions.h`. We'd have to deal with the additional complexity of registering in both locations. Also for the normal non-mobile case we'd still need to handle cuda registrations in separately, so we'd have registrations in 3 places. Personally I don't think this if-statement dispatch is very complex in comparison to that.

smessmer replied:
ok let's go with the current approach for now, we'll probably have to change this later.

smessmer commented on c10/core/TensorImpl.cpp:
```diff
@@ -196,7 +196,11 @@ std::unique_ptr<NamedTensorMetaInterface> NamedTensorMetaInterface::clone() cons
 #ifndef CAFFE2_FB_LIMITED_MOBILE_CAPABILITY
 
+#ifdef USE_STATIC_DISPATCH
+thread_local bool NonVariableTypeMode_enabled = true;
```
Why is variable type mode different depending on static dispatch?

li-roy replied:
For mobile, we're never going through Variable code, but we're passing Variable code through TH methods that expect tensor. Specifically `checked_tensor_unwrap` does an `is_variable()` check, and it'll fail because NonVariableTypeMode is always off, even though we are never going through Variable code. Because our constraints are inference-only and never going through VariableType, I thought it would make sense to just set NonVariableTypeMode and keep it.

@yf225 Do you have any thoughts on this?

ljk53 replied:
After @yf225's Variable/Tensor unification work, do we still need keep is_variable() check?

Earlier (before Will landed his work) when I tried removing virtual methods on types & variable, I figured it mostly worked fine without is_variable() check, only need keep a few variable overridden methods virtual:

https://github.com/ljk53/pytorch/commit/2f795a50715f014737d0397f76ed11605939124a
https://github.com/ljk53/pytorch/commit/cc8a1787f79e77cfcff759f92f226afd09feb09a

It's not needed by mobile inference for now but we are still discussing whether we need variable/autodiff for federate on mobile in the future.

dzhulgakov replied:
It's a bit scary - I don't say we will add autograd on mobile build, but coupling with static dispatch like that is suspicious. Can't we just have a lightweight guard (at::AutoNonVariableTypeMode) in every method call?

Probably even better would be to just make sure we don't produce Variables at all - e.g. stub out factory functions if autograd is not compiled in (do we have a dedicated flag for turning off autograd?)

smessmer reviewed (changes requested):
There's a few questions we should talk about before landing this, but generally looks good.

li-roy commented on aten/src/ATen/ScalarOps.h:
```diff
@@ -2,6 +2,7 @@
 
 #include <c10/core/Scalar.h>
 #include <ATen/Tensor.h>
+#include <ATen/Functions.h>
```
needed for `at::scalar_tensor`

li-roy commented on CMakeLists.txt:
```diff
@@ -431,6 +436,7 @@ include_directories(BEFORE ${PROJECT_SOURCE_DIR})
 include_directories(BEFORE ${PROJECT_BINARY_DIR})
 
 include_directories(BEFORE ${PROJECT_SOURCE_DIR}/aten/src/)
+include_directories(BEFORE ${PROJECT_BINARY_DIR}/aten/src/)
```
We need to link generated ATen (non-core) in addition to non-generated. Without this, when we build `TensorMethods.h` in mobile, stuff like `TypeDefault.h` won't have been generated yet.

li-roy commented on .circleci/config.yml:
```diff
@@ -95,6 +95,11 @@ pytorch_linux_build_defaults: &pytorch_linux_build_defaults
 
+        # dispatch aten ops statically for mobile
+        if [[ ${BUILD_ENVIRONMENT} == *"android"* ]]; then
+          NAMED_FLAG="export USE_STATIC_DISPATCH=1"
```
Yeah I think what we do for this depends on what we decide to do for specifying the subset of ops. Can we move forward with a boolean flag for now and change it later if we need to? We can change the name if you prefer, or maybe just pass an existing flag. But as far as I know, there's not a single existing flag that does what we want, because we don't want this to be triggered for internal mobile builds. @ljk53 any thoughts?

ljk53 replied:
I lean towards having separate flags for now. We might still need decide how to specify ops whitelist, e.g. what if we want to choose between ops having same name? what if we want to use config file instead of encoded string? We can always make the "static dispatch" flag "intern_" flag and set it automatically later (IMO keeping it as separate intern flag is easier to understand anyway).

BTW, you probably only need modify .jenkins/pytorch/build.sh (see how DBUILD_CAFFE2_MOBILE is set there).

ajtulloch replied:
Is the plan for PyTorch Mobile to like, completely ignore OpenGL, Metal, Vulkan, etc?

ljk53 replied:
@ajtulloch when we get there can we use switch-case or hashmap-of-function-pointers approach? The first step is to get rid of huge aten vtable which blocks linker from striping out unused code...

dzhulgakov replied:
Static dispatch doesn't imply only CPU - the code below already generates the switch statements (which is good). We should add a configurable filter on dispatch ids (basically selecting which devices to compile), but it can be done separately.

Commit:
Update on "Add static dispatch mode to reduce mobile code size"

Add static dispatch mode to reduce mobile code size

gh-metadata: pytorch pytorch 22335 gh/li-roy/36/head

Commit:
Update on "Add static dispatch mode to reduce mobile code size"

Add static dispatch mode to reduce mobile code size

gh-metadata: pytorch pytorch 22335 gh/li-roy/36/head

dzhulgakov commented on .circleci/verbatim-sources/linux-build-defaults.yml:
```diff
@@ -37,6 +37,11 @@ pytorch_linux_build_defaults: &pytorch_linux_build_defaults
         fi
 
+        # dispatch aten ops statically for mobile
+        if [[ ${BUILD_ENVIRONMENT} == *"android"* ]]; then
```
we should also filter for iOS here, @xta0 - what is the right way to do it?

li-roy replied:
I think we don't have ios CI yet.

li-roy commented:
@pytorchbot retest this please

Commit:
Update on "Add static dispatch mode to reduce mobile code size"

Add static dispatch mode to reduce mobile code size

gh-metadata: pytorch pytorch 22335 gh/li-roy/36/head

Commit:
Update on "Add static dispatch mode to reduce mobile code size"

Add static dispatch mode to reduce mobile code size

gh-metadata: pytorch pytorch 22335 gh/li-roy/36/head

smessmer commented:
looks good, thanks

dzhulgakov commented on tools/jit/gen_jit_dispatch.py:
```diff
@@ -144,7 +144,11 @@ def from_ivalue(arg, value):
         .device(${device})
         .pinned_memory(${pin_memory});
-auto result_ = torch::${name}(${args_with_tensor_options});
+#ifdef USE_STATIC_DISPATCH
```
Is it to avoid creating variables? It might be better to put NoGrad guard on. Afaiu, we want to get rid of at:: factory functions eventually and just create always variables (cc @gchanan)

li-roy replied:
Yeah it's to avoid creating variables. I don't think NoGrad works without additional changes, torch:: will always create a variable.

ljk53 replied:
I plan to look into optional build for autograd/variable functionality on top of this PR and make changes if I find anything - so it's fine as long as the static dispatching part works.

ljk53 commented on aten/src/ATen/function_wrapper.py:
```diff
@@ -131,8 +131,12 @@ def TypedDict(name, attrs, total=True):  # type: ignore
 TENSOR_METHOD_DEFINITION = CodeTemplate("""\
 inline ${return_type} Tensor::${api_name}(${method_formals}) const {
+#ifdef USE_STATIC_DISPATCH
+    ${mobile_method_body}
```
nit: you might want to call this "static_dispatch_method_body" to be consistent with the macro name?

ljk53 commented on aten/src/ATen/function_wrapper.py:
```diff
@@ -205,6 +236,8 @@ def TypedDict(name, attrs, total=True):  # type: ignore
     ('BFloat16', 'BFloat16', 'BFloat16AccrealNotDefined', True),
 ]
 
+mobile_backends = ['CPU', 'QuantizedCPU', 'SparseCPU']
```
static_dispatch_backends?

ezyang commented:
This broke the named tensor master-only build. @zou3519 https://circleci.com/gh/pytorch/pytorch/2519587?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link/console
